File: ios/Heritg/HeritgZIP.swift
import Foundation

nonisolated enum HeritgZIP {
    struct Entry: Sendable {
        let path: String
        let data: Data
    }

    private struct CentralEntry {
        let path: String
        let crc32: UInt32
        let size: Int
        let localOffset: Int
    }

    private static let crcTable: [UInt32] = (0..<256).map { index in
        var value = UInt32(index)
        for _ in 0..<8 { value = (value >> 1) ^ (value & 1 == 1 ? 0xedb88320 : 0) }
        return value
    }

    static func encode(_ entries: [Entry]) throws -> Data {
        guard entries.count <= Int(UInt16.max) else { throw HeritgArchiveError.tooManyRecords }
        let sorted = entries.sorted { $0.path < $1.path }
        guard Set(sorted.map(\.path)).count == sorted.count else {
            throw HeritgArchiveError.invalidArchive
        }

        var archive = Data()
        var centralRecords = [(entry: Entry, crc: UInt32, offset: UInt32)]()
        for entry in sorted {
            try validate(path: entry.path)
            guard entry.data.count <= Int(UInt32.max), archive.count <= Int(UInt32.max) else {
                throw HeritgArchiveError.fileTooLarge
            }
            let name = Data(entry.path.utf8)
            guard name.count <= Int(UInt16.max) else { throw HeritgArchiveError.fieldTooLarge }
            let crc = crc32(entry.data)
            centralRecords.append((entry, crc, UInt32(archive.count)))
            archive.appendZIPUInt32(0x04034b50)
            archive.appendZIPUInt16(20)
            archive.appendZIPUInt16(0x0800)
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt16(0x0021)
            archive.appendZIPUInt32(crc)
            archive.appendZIPUInt32(UInt32(entry.data.count))
            archive.appendZIPUInt32(UInt32(entry.data.count))
            archive.appendZIPUInt16(UInt16(name.count))
            archive.appendZIPUInt16(0)
            archive.append(name)
            archive.append(entry.data)
        }

        guard archive.count <= Int(UInt32.max) else { throw HeritgArchiveError.fileTooLarge }
        let centralOffset = UInt32(archive.count)
        for record in centralRecords {
            let name = Data(record.entry.path.utf8)
            archive.appendZIPUInt32(0x02014b50)
            archive.appendZIPUInt16(20)
            archive.appendZIPUInt16(20)
            archive.appendZIPUInt16(0x0800)
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt16(0x0021)
            archive.appendZIPUInt32(record.crc)
            archive.appendZIPUInt32(UInt32(record.entry.data.count))
            archive.appendZIPUInt32(UInt32(record.entry.data.count))
            archive.appendZIPUInt16(UInt16(name.count))
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt16(0)
            archive.appendZIPUInt32(0)
            archive.appendZIPUInt32(record.offset)
            archive.append(name)
        }
        let centralSize = archive.count - Int(centralOffset)
        guard centralSize <= Int(UInt32.max) else { throw HeritgArchiveError.fileTooLarge }
        archive.appendZIPUInt32(0x06054b50)
        archive.appendZIPUInt16(0)
        archive.appendZIPUInt16(0)
        archive.appendZIPUInt16(UInt16(centralRecords.count))
        archive.appendZIPUInt16(UInt16(centralRecords.count))
        archive.appendZIPUInt32(UInt32(centralSize))
        archive.appendZIPUInt32(centralOffset)
        archive.appendZIPUInt16(0)
        guard archive.count <= HeritgArchive.maximumFileBytes else {
            throw HeritgArchiveError.fileTooLarge
        }
        return archive
    }

    static func decode(_ archive: Data) throws -> [String: Data] {
        guard archive.count <= HeritgArchive.maximumFileBytes else {
            throw HeritgArchiveError.fileTooLarge
        }
        guard archive.count >= 22 else { throw HeritgArchiveError.invalidArchive }
        let endOffset = archive.count - 22
        guard try archive.zipUInt32(at: endOffset) == 0x06054b50,
              try archive.zipUInt16(at: endOffset + 4) == 0,
              try archive.zipUInt16(at: endOffset + 6) == 0,
              try archive.zipUInt16(at: endOffset + 8) == archive.zipUInt16(at: endOffset + 10),
              try archive.zipUInt16(at: endOffset + 20) == 0 else {
            throw HeritgArchiveError.invalidArchive
        }
        let entryCount = Int(try archive.zipUInt16(at: endOffset + 10))
        let centralSize = Int(try archive.zipUInt32(at: endOffset + 12))
        let centralOffset = Int(try archive.zipUInt32(at: endOffset + 16))
        guard centralOffset + centralSize == endOffset else {
            throw HeritgArchiveError.invalidArchive
        }

        var centralEntries = [CentralEntry]()
        var cursor = centralOffset
        var names = Set<String>()
        for _ in 0..<entryCount {
            guard try archive.zipUInt32(at: cursor) == 0x02014b50 else {
                throw HeritgArchiveError.invalidArchive
            }
            let madeBy = try archive.zipUInt16(at: cursor + 4)
            let flags = try archive.zipUInt16(at: cursor + 8)
            let method = try archive.zipUInt16(at: cursor + 10)
            let crc = try archive.zipUInt32(at: cursor + 16)
            let compressedSize = Int(try archive.zipUInt32(at: cursor + 20))
            let size = Int(try archive.zipUInt32(at: cursor + 24))
            let nameLength = Int(try archive.zipUInt16(at: cursor + 28))
            let extraLength = Int(try archive.zipUInt16(at: cursor + 30))
            let commentLength = Int(try archive.zipUInt16(at: cursor + 32))
            let disk = try archive.zipUInt16(at: cursor + 34)
            let externalAttributes = try archive.zipUInt32(at: cursor + 38)
            let localOffset = Int(try archive.zipUInt32(at: cursor + 42))
            guard flags == 0x0800, method == 0, compressedSize == size,
                  extraLength == 0, commentLength == 0, disk == 0 else {
                throw HeritgArchiveError.invalidArchive
            }
            if madeBy >> 8 == 3 {
                let fileType = (externalAttributes >> 16) & 0xf000
                guard fileType == 0 || fileType == 0x8000 else {
                    throw HeritgArchiveError.invalidArchive
                }
            }
            let nameStart = cursor + 46
            let nameData = try archive.zipData(in: nameStart..<(nameStart + nameLength))
            guard let path = String(data: nameData, encoding: .utf8) else {
                throw HeritgArchiveError.invalidArchive
            }
            try validate(path: path)
            guard names.insert(path).inserted else { throw HeritgArchiveError.invalidArchive }
            centralEntries.append(CentralEntry(path: path, crc32: crc, size: size, localOffset: localOffset))
            cursor = nameStart + nameLength
        }
        guard cursor == endOffset else { throw HeritgArchiveError.invalidArchive }

        var result = [String: Data]()
        var localRanges = [Range<Int>]()
        var totalSize = 0
        for entry in centralEntries {
            let offset = entry.localOffset
            guard try archive.zipUInt32(at: offset) == 0x04034b50,
                  try archive.zipUInt16(at: offset + 6) == 0x0800,
                  try archive.zipUInt16(at: offset + 8) == 0,
                  try archive.zipUInt32(at: offset + 14) == entry.crc32,
                  Int(try archive.zipUInt32(at: offset + 18)) == entry.size,
                  Int(try archive.zipUInt32(at: offset + 22)) == entry.size else {
                throw HeritgArchiveError.invalidArchive
            }
            let nameLength = Int(try archive.zipUInt16(at: offset + 26))
            let extraLength = Int(try archive.zipUInt16(at: offset + 28))
            guard extraLength == 0 else { throw HeritgArchiveError.invalidArchive }
            let nameStart = offset + 30
            let dataStart = nameStart + nameLength
            let end = dataStart + entry.size
            guard end <= centralOffset,
                  String(data: try archive.zipData(in: nameStart..<dataStart), encoding: .utf8) == entry.path else {
                throw HeritgArchiveError.invalidArchive
            }
            let data = try archive.zipData(in: dataStart..<end)
            guard crc32(data) == entry.crc32 else { throw HeritgArchiveError.invalidArchive }
            totalSize += data.count
            guard totalSize <= HeritgArchive.maximumFileBytes else {
                throw HeritgArchiveError.fileTooLarge
            }
            localRanges.append(offset..<end)
            result[entry.path] = data
        }
        var expectedStart = 0
        for range in localRanges.sorted(by: { $0.lowerBound < $1.lowerBound }) {
            guard range.lowerBound == expectedStart else { throw HeritgArchiveError.invalidArchive }
            expectedStart = range.upperBound
        }
        guard expectedStart == centralOffset else { throw HeritgArchiveError.invalidArchive }
        return result
    }

    private static func validate(path: String) throws {
        guard !path.isEmpty,
              path.utf8.count <= HeritgArchive.maximumShortFieldBytes,
              !path.hasPrefix("/"), !path.hasSuffix("/"), !path.contains("\\"),
              !path.unicodeScalars.contains(where: { $0.value == 0 }),
              path.split(separator: "/", omittingEmptySubsequences: false).allSatisfy({ $0 != "." && $0 != ".." && !$0.isEmpty }) else {
            throw HeritgArchiveError.invalidArchive
        }
    }

    private static func crc32(_ data: Data) -> UInt32 {
        var crc = UInt32.max
        for byte in data {
            crc = (crc >> 8) ^ crcTable[Int((crc ^ UInt32(byte)) & 0xff)]
        }
        return crc ^ UInt32.max
    }
}

private extension Data {
    mutating func appendZIPUInt16(_ value: UInt16) {
        append(UInt8(truncatingIfNeeded: value))
        append(UInt8(truncatingIfNeeded: value >> 8))
    }

    mutating func appendZIPUInt32(_ value: UInt32) {
        appendZIPUInt16(UInt16(truncatingIfNeeded: value))
        appendZIPUInt16(UInt16(truncatingIfNeeded: value >> 16))
    }

    func zipUInt16(at offset: Int) throws -> UInt16 {
        guard offset >= 0, offset <= count - 2 else { throw HeritgArchiveError.invalidArchive }
        return UInt16(self[offset]) | (UInt16(self[offset + 1]) << 8)
    }

    func zipUInt32(at offset: Int) throws -> UInt32 {
        guard offset >= 0, offset <= count - 4 else { throw HeritgArchiveError.invalidArchive }
        return UInt32(self[offset])
            | (UInt32(self[offset + 1]) << 8)
            | (UInt32(self[offset + 2]) << 16)
            | (UInt32(self[offset + 3]) << 24)
    }

    func zipData(in range: Range<Int>) throws -> Data {
        guard range.lowerBound >= 0, range.upperBound >= range.lowerBound,
              range.upperBound <= count else {
            throw HeritgArchiveError.invalidArchive
        }
        return subdata(in: range)
    }
}

File: android/core/src/main/java/tech/robihamanto/heritg/android/core/interop/HeritgZip.kt
package tech.robihamanto.heritg.android.core.interop

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.util.zip.CRC32

internal object HeritgZip {
    private const val LocalSignature = 0x04034b50L
    private const val CentralSignature = 0x02014b50L
    private const val EndSignature = 0x06054b50L
    private const val Version = 20
    private const val Flags = 0x0800
    private const val Stored = 0
    private const val DosTime = 0
    private const val DosDate = 0x0021

    private data class Central(
        val path: String,
        val crc: Long,
        val size: Int,
        val localOffset: Int,
    )

    fun encode(entries: Map<String, ByteArray>): ByteArray {
        if (entries.size > UShort.MAX_VALUE.toInt()) throw ArchiveException.TooManyRecords()
        val sorted = entries.entries.sortedBy { it.key }
        val output = ByteArrayOutputStream()
        val central = mutableListOf<Central>()
        sorted.forEach { (path, data) ->
            validatePath(path)
            val name = path.encodeToByteArray()
            if (name.size > UShort.MAX_VALUE.toInt()) throw ArchiveException.FieldTooLarge()
            val crc = crc32(data)
            central += Central(path, crc, data.size, output.size())
            output.u32(LocalSignature)
            output.u16(Version)
            output.u16(Flags)
            output.u16(Stored)
            output.u16(DosTime)
            output.u16(DosDate)
            output.u32(crc)
            output.u32(data.size.toLong())
            output.u32(data.size.toLong())
            output.u16(name.size)
            output.u16(0)
            output.write(name)
            output.write(data)
            checkSize(output.size())
        }
        val centralOffset = output.size()
        central.forEach { record ->
            val name = record.path.encodeToByteArray()
            output.u32(CentralSignature)
            output.u16(Version)
            output.u16(Version)
            output.u16(Flags)
            output.u16(Stored)
            output.u16(DosTime)
            output.u16(DosDate)
            output.u32(record.crc)
            output.u32(record.size.toLong())
            output.u32(record.size.toLong())
            output.u16(name.size)
            output.u16(0)
            output.u16(0)
            output.u16(0)
            output.u16(0)
            output.u32(0)
            output.u32(record.localOffset.toLong())
            output.write(name)
        }
        val centralSize = output.size() - centralOffset
        output.u32(EndSignature)
        output.u16(0)
        output.u16(0)
        output.u16(central.size)
        output.u16(central.size)
        output.u32(centralSize.toLong())
        output.u32(centralOffset.toLong())
        output.u16(0)
        checkSize(output.size())
        return output.toByteArray()
    }

    fun decode(archive: ByteArray): Map<String, ByteArray> {
        if (archive.size > ArchiveConstants.MaximumArchiveBytes) throw ArchiveException.FileTooLarge()
        if (archive.size < 22) throw ArchiveException.InvalidArchive()
        val end = archive.size - 22
        if (archive.u32(end) != EndSignature || archive.u16(end + 4) != 0 || archive.u16(end + 6) != 0 ||
            archive.u16(end + 8) != archive.u16(end + 10) || archive.u16(end + 20) != 0
        ) throw ArchiveException.InvalidArchive()
        val entryCount = archive.u16(end + 10)
        val centralSize = archive.u32(end + 12).toIntExact()
        val centralOffset = archive.u32(end + 16).toIntExact()
        if (centralOffset + centralSize != end) throw ArchiveException.InvalidArchive()

        val central = mutableListOf<Central>()
        val names = mutableSetOf<String>()
        var cursor = centralOffset
        repeat(entryCount) {
            if (archive.u32(cursor) != CentralSignature) throw ArchiveException.InvalidArchive()
            val madeBy = archive.u16(cursor + 4)
            val needed = archive.u16(cursor + 6)
            val flags = archive.u16(cursor + 8)
            val method = archive.u16(cursor + 10)
            val time = archive.u16(cursor + 12)
            val date = archive.u16(cursor + 14)
            val crc = archive.u32(cursor + 16)
            val compressed = archive.u32(cursor + 20).toIntExact()
            val size = archive.u32(cursor + 24).toIntExact()
            val nameLength = archive.u16(cursor + 28)
            val extraLength = archive.u16(cursor + 30)
            val commentLength = archive.u16(cursor + 32)
            val disk = archive.u16(cursor + 34)
            val internalAttributes = archive.u16(cursor + 36)
            val externalAttributes = archive.u32(cursor + 38)
            val localOffset = archive.u32(cursor + 42).toIntExact()
            if (needed != Version || flags != Flags || method != Stored || time != DosTime || date != DosDate ||
                compressed != size || extraLength != 0 || commentLength != 0 || disk != 0 || internalAttributes != 0
            ) throw ArchiveException.InvalidArchive()
            if (madeBy ushr 8 == 3) {
                val fileType = (externalAttributes ushr 16) and 0xf000
                if (fileType != 0L && fileType != 0x8000L) throw ArchiveException.InvalidArchive()
            }
            val nameStart = cursor + 46
            val path = archive.utf8(nameStart, nameStart + nameLength)
            validatePath(path)
            if (!names.add(path)) throw ArchiveException.InvalidArchive()
            central += Central(path, crc, size, localOffset)
            cursor = nameStart + nameLength
        }
        if (cursor != end) throw ArchiveException.InvalidArchive()

        val result = linkedMapOf<String, ByteArray>()
        val ranges = mutableListOf<IntRange>()
        var totalSize = 0L
        central.forEach { entry ->
            val offset = entry.localOffset
            if (archive.u32(offset) != LocalSignature || archive.u16(offset + 4) != Version ||
                archive.u16(offset + 6) != Flags || archive.u16(offset + 8) != Stored ||
                archive.u16(offset + 10) != DosTime || archive.u16(offset + 12) != DosDate ||
                archive.u32(offset + 14) != entry.crc || archive.u32(offset + 18).toIntExact() != entry.size ||
                archive.u32(offset + 22).toIntExact() != entry.size
            ) throw ArchiveException.InvalidArchive()
            val nameLength = archive.u16(offset + 26)
            val extraLength = archive.u16(offset + 28)
            if (extraLength != 0) throw ArchiveException.InvalidArchive()
            val nameStart = offset + 30
            val dataStart = nameStart + nameLength
            val dataEnd = dataStart + entry.size
            if (dataEnd > centralOffset || archive.utf8(nameStart, dataStart) != entry.path) {
                throw ArchiveException.InvalidArchive()
            }
            val data = archive.slice(dataStart, dataEnd)
            if (crc32(data) != entry.crc) throw ArchiveException.InvalidArchive()
            totalSize += data.size
            if (totalSize > ArchiveConstants.MaximumArchiveBytes) throw ArchiveException.FileTooLarge()
            ranges += offset until dataEnd
            result[entry.path] = data
        }
        var expectedStart = 0
        ranges.sortedBy { it.first }.forEach {
            if (it.first != expectedStart) throw ArchiveException.InvalidArchive()
            expectedStart = it.last + 1
        }
        if (expectedStart != centralOffset) throw ArchiveException.InvalidArchive()
        return result
    }

    fun validatePath(path: String) {
        val components = path.split('/', ignoreCase = false, limit = 0)
        if (path.isEmpty() || path.encodeToByteArray().size > ArchiveConstants.MaximumShortFieldBytes ||
            path.startsWith('/') || path.endsWith('/') || '\\' in path || '\u0000' in path ||
            components.any { it.isEmpty() || it == "." || it == ".." }
        ) throw ArchiveException.InvalidArchive()
    }

    private fun checkSize(size: Int) {
        if (size > ArchiveConstants.MaximumArchiveBytes) throw ArchiveException.FileTooLarge()
    }

    private fun crc32(bytes: ByteArray): Long = CRC32().apply { update(bytes) }.value

    private fun ByteArray.u16(offset: Int): Int {
        if (offset < 0 || offset > size - 2) throw ArchiveException.InvalidArchive()
        return (this[offset].toInt() and 0xff) or ((this[offset + 1].toInt() and 0xff) shl 8)
    }

    private fun ByteArray.u32(offset: Int): Long {
        if (offset < 0 || offset > size - 4) throw ArchiveException.InvalidArchive()
        return (this[offset].toLong() and 0xff) or
            ((this[offset + 1].toLong() and 0xff) shl 8) or
            ((this[offset + 2].toLong() and 0xff) shl 16) or
            ((this[offset + 3].toLong() and 0xff) shl 24)
    }

    private fun ByteArray.slice(start: Int, end: Int): ByteArray {
        if (start < 0 || end < start || end > size) throw ArchiveException.InvalidArchive()
        return copyOfRange(start, end)
    }

    private fun ByteArray.utf8(start: Int, end: Int): String = try {
        StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(slice(start, end))).toString()
    } catch (error: Exception) {
        throw ArchiveException.InvalidArchive(error)
    }

    private fun Long.toIntExact(): Int =
        if (this > Int.MAX_VALUE) throw ArchiveException.InvalidArchive() else toInt()

    private fun ByteArrayOutputStream.u16(value: Int) {
        write(value and 0xff)
        write((value ushr 8) and 0xff)
    }

    private fun ByteArrayOutputStream.u32(value: Long) {
        u16((value and 0xffff).toInt())
        u16(((value ushr 16) and 0xffff).toInt())
    }
}
